base/math.jl

```diff
@@ -70,11 +71,23 @@ end
 
 # evaluate p[1] + x * (p[2] + x * (....)), i.e. a polynomial via Horner's rule
 macro horner(x, p...)
+    @gensym val
     ex = esc(p[end])
     for i = length(p)-1:-1:1
-        ex = :(muladd(t, $ex, $(esc(p[i]))))
+        ex = :(muladd($val, $ex, $(esc(p[i]))))
     end
-    Expr(:block, :(t = $(esc(x))), ex)
+    Expr(:block, :($val = $(esc(x))), ex)
+end
+
+# evaluate p[1] + x * (p[2] + x * (....)), i.e. a polynomial via Horner's rule
+# and convert coefficients to same type as x
+macro horner_oftype(x, p...)
```

Comment:
> I would rather not add the extra macro for an odd case (it is unusual that you wouldn't be able to reduce the degree for Float16) that we don't really implement 

Not really unusual for Float16 (as opposed to going from  Float128 -> Float64). You have a lot less precision so it's very hard to find an even smaller degree polynomial that retains the same accuracy constraint ( < 1 ulp). This is the hardest/time consuming part testing accuracy and the polynomial coeffs.

>  (probably` on CPUs we will just compute exp(x::Float16) by converting to/from Float32, no).

yes

I will remove anyways